lifecycle-hooks.md

```diff
@@ -0,0 +1,188 @@
+
+Firstly we will need a simple extension server that needs to run and be able to receive cluster lifecycle events. For this, lets create a deployment in the management cluster that will act as our extension server.
+
+```bash
+kubectl apply -f yamls/extension/test-extension-deployment.yaml
+```
+
+Note that the YAML also contains all necessary resources like Service, Certificates, etc that will make this extension server reachable from within the management cluster.
+
+Let's verify that the extension server is running:
+
+```bash
+kubectl get deployments -n test-extension-system
+```
+
+The output should resemble:
+
+```bash
+NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
+test-extension-manager   1/1     1            1           30s
+```
+
+## Register the Extension
+
+Now that we have a running extension server let's register it with Cluster API. This will inform Cluster API to call the extension server during key workload cluster lifecycle events.
+
+Let's first look at how the registration looks like:
+
+```yaml
+apiVersion: runtime.cluster.x-k8s.io/v1alpha1
+kind: ExtensionConfig
+metadata:
+  name: printer-extension                      // Unique Registration Name
+  annotations:
+    runtime.cluster.x-k8s.io/inject-ca-from-secret: test-extension-system/test-extension-webhook-service-cert
+spec:
+  clientConfig:                                // Information to contact the extension server
+    service:                                   // Reference to a service running in the management cluster
+      name: test-extension-webhook-service
+      namespace: test-extension-system
+```
+
+Let's register the extension using:
+
+```bash
+kubectl apply -f yamls/extension/test-extension-config.yaml
+```
+
+Let us now verify that the extension is properly registered and reachable:
+
+```bash
+kubectl get extensionconfig printer-extension -o jsonpath='{.status.conditions}'
+```
+
+The output should resemble:
+
+```bash
+[{"lastTransitionTime":"2022-10-12T04:55:44Z","status":"True","type":"Discovered"}]
+```
+
+> **Optional**  
+> If you are curious, take a look at `.status.handlers` of the `printer-extension` resource by doing `kubectl get extensionconfig printer-extension -o jsonpath='{.status.handlers}'`. This lists down all the lifecycle events the extension server would like to receive.
+
+## Extension in Action
+
+Now that we have the extension server running and registered with Cluster API let's see it in action. The extension server is configured to receive the following events (non-exhaustive list):
+
+- BeforeClusterCreate
+- AfterControlPlaneInitialized
+- BeforeClusterDelete
+
+### Create a new workload Cluster
+
+In this section we will create a new workload cluster and see that the extension server receives these events and logs them.  
+In a new terminal tab let's start watching the logs of the extension server:
+
+```bash
+kubectl logs -n test-extension-system deployments/test-extension-manager -f
```

Comment:
Maybe offer the non-follow logs command as an option here (after creating the cluster)

Reply:
Is done